File: 2019/day03_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Coords = tuple[int, int]
type Instruction = tuple[str, int]
type Segment = tuple[Coords, Coords]


def solution(content: str) -> Any:
    lines = content.split("\n")
    return sorted(
        [
            manhattan((0, 0), coords)
            for s1, s2 in [
                (s1, s2)
                for s1 in parse(lines[0])
                for s2 in parse(lines[1])
                if intersect(s1, s2)
            ]
            for coords in unfold(s1)
            if coords in unfold(s2) and coords != (0, 0)
        ]
    )[0]


def unfold(segment: Segment):
    start, end = segment
    if is_vertical(segment) and start[1] < end[1]:
        s, e = (start[1], end[1]) if start[1] < end[1] else (end[1], start[1])
        return [(start[0], y) for y in range(s, e + 1)]
    else:
        s, e = (start[0], end[0]) if start[0] < end[0] else (end[0], start[0])
        return [(x, start[1]) for x in range(s, e + 1)]


def intersect(s1: Segment, s2: Segment):
    start1, end1 = s1
    start2, end2 = s2
    if is_vertical(s1) and is_vertical(s2):
        return start1[1] == start2[1] and (
            between(start1[1], start2[1], end2[1])
            or between(end1[1], start2[1], end2[1])
            or between(start2[1], start1[1], end1[1])
            or between(end2[1], start1[1], end1[1])
        )
    if not is_vertical(s1) and is_vertical(s2):
        return between(start1[1], start2[1], end2[1]) and between(
            start2[0], start1[0], end1[0]
        )
    if is_vertical(s1) and not is_vertical(s2):
        return between(start1[0], start2[0], end2[0]) and between(
            start2[1], start1[1], end1[1]
        )
    if not is_vertical(s1) and not is_vertical(s2):
        return start1[0] == start2[0] and (
            (
                between(start1[0], start2[0], end2[0])
                or between(end1[0], start2[0], end2[0])
                or between(start2[0], start1[0], end1[0])
                or between(end2[0], start1[0], end1[0])
            )
        )
    return False


def is_vertical(seg: Segment):
    start, end = seg
    return start[0] == end[0]


def between(x: int, start: int, end: int):
    return (start <= x <= end) if start < end else (end <= x <= start)


def parse(line: str) -> list[Segment]:
    segments = []
    for instruction in [parse_instruction(word) for word in line.split(",")]:
        if segments:
            _, last = segments[-1]
            segments.append((last, move(last, instruction)))
        else:
            segments.append(((0, 0), move((0, 0), instruction)))
    return segments


def move(start: Coords, instr: Instruction):
    if instr[0] == "R":
        return (start[0] + instr[1], start[1])
    if instr[0] == "L":
        return (start[0] - instr[1], start[1])
    if instr[0] == "U":
        return (start[0], start[1] + instr[1])
    if instr[0] == "D":
        return (start[0], start[1] - instr[1])
    return start


def parse_instruction(word: str) -> Instruction:
    return (word[0], int(word[1:]))


def manhattan(c1: Coords, c2: Coords):
    return abs(c1[0] - c2[0]) + abs(c1[1] - c2[1])


def main() -> None:
    with open("day03.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2019/day03-1.hs
import Data.List (sort)

data Direction = U | D | L | R

type Coords = (Int, Int)

type Instruction = (Direction, Int)

type Segment = (Coords, Coords)

solution :: String -> Int
solution content =
  minimum . map (manhattan (0, 0)) $
    [ coords
      | s1 <- w1Seg,
        s2 <- w2Seg,
        s1 `intersect` s2,
        coords <- Main.unfold s1,
        coords `elem` Main.unfold s2,
        coords /= (0, 0)
    ]
  where
    w1Seg = parse . (!! 0) . lines $ content
    w2Seg = parse . (!! 1) . lines $ content
    wire1 = concatMap unfold w1Seg
    wire2 = concatMap unfold w2Seg

unfold :: Segment -> [Coords]
unfold ((sx, sy), (ex, ey))
  | isVertical ((sx, sy), (ex, ey)) = [(sx, y) | y <- [(min sy ey) .. (max sy ey)]]
  | otherwise = [(x, sy) | x <- [(min sx ex) .. (max sx ex)]]

intersect :: Segment -> Segment -> Bool
intersect ((s1x, s1y), (e1x, e1y)) ((s2x, s2y), (e2x, e2y))
  | isVertical ((s1x, s1y), (e1x, e1y)) && isVertical ((s2x, s2y), (e2x, e2y)) =
      (s1y == s2y)
        && ( between s1y s2y e2y
               || between e1y s2y e2y
               || between s2y s1y e1y
               || between e2y s1y e1y
           )
  | not (isVertical ((s1x, s1y), (e1x, e1y))) && isVertical ((s2x, s2y), (e2x, e2y)) = Main.between s1y s2y e2y && Main.between s2x s1x e1x
  | isVertical ((s1x, s1y), (e1x, e1y)) && not (isVertical ((s2x, s2y), (e2x, e2y))) = Main.between s1x s2x e2x && Main.between s2y s1y e1y
  | otherwise =
      (s1x == s2x)
        && ( between s1x s2x e2x
               || between e1x s2x e2x
               || between s2x s1x e1x
               || between e2x s1x e1x
           )

isVertical :: Segment -> Bool
isVertical ((sx, _), (ex, _)) = sx == ex

between :: (Ord a) => a -> a -> a -> Bool
between x start end
  | start < end = start <= x && x <= end
  | otherwise = end <= x && x <= start

parse :: String -> [Segment]
parse line = helper [] . map parseInstruction $ splitOn ',' line
  where
    helper [] (instr : xs) = ((0, 0), move (0, 0) instr) : helper [((0, 0), move (0, 0) instr)] xs
    helper segments [] = segments
    helper segments (instr : xs) = segments ++ [(snd . last $ segments, move (snd . last $ segments) instr)] ++ helper [(snd . last $ segments, move (snd . last $ segments) instr)] xs

splitOn :: Char -> String -> [String]
splitOn c xs = helper [""] c xs
  where
    helper o c "" = o
    helper o c (x : "")
      | x == c = o
      | otherwise = init o ++ [last o ++ [x]]
    helper o c (x : xs)
      | x == c = helper (o ++ [""]) c xs
      | otherwise = helper (init o ++ [last o ++ [x]]) c xs

move :: Coords -> Instruction -> Coords
move (x, y) (R, amount) = (x + amount, y)
move (x, y) (L, amount) = (x - amount, y)
move (x, y) (U, amount) = (x, y + amount)
move (x, y) (D, amount) = (x, y - amount)

parseInstruction :: String -> Instruction
parseInstruction ('U' : xs) = (U, read xs)
parseInstruction ('D' : xs) = (D, read xs)
parseInstruction ('L' : xs) = (L, read xs)
parseInstruction ('R' : xs) = (R, read xs)

manhattan :: (Num a) => (a, a) -> (a, a) -> a
manhattan (x1, y1) (x2, y2) = abs (x1 - x2) + abs (y1 - y2)

main :: IO ()
main = do readFile "day03.dat" >>= print . solution
